experimental/mrpc/plugin/mrpc/src/engine.rs

```diff
@@ -478,18 +482,49 @@ impl MrpcEngine {
                             shm_addr_backend: msg.addr_backend,
                         };
                         // timer.tick();
+                        match meta.status_code {
+                            StatusCode::AccessDenied => {
+                                tracing::trace!("Status code: Access denied, meta={:?}", meta);
+                                let mut sent = false;
+                                let rpc_id = RpcId(meta.conn_id, meta.call_id);
+                                let status = phoenix_api::rpc::TransportStatus::Error(unsafe {
+                                    NonZeroU32::new_unchecked(402)
```

Comment:
I am not sure about how to do that. Currently, I have added `StatusCode` in meta to get the error code. 
However, I cannot directly bubble status code to the actual message that client sees, but must convert it to `TransportStatus`. But that is a little bit cubersome, and awkward, since I believe  `TransportStatus` should be used to indicate error in transport layer, and I don't think it is a good idea to mix those status code.

One possible solution is to replace `StatusCode` with `TransportStatus` in meta, with a different name, maybe, and then we can use the enums to get meaningful error code. Another question is whether the current implementation of `TransportStatus` can be safely passed through network, since it uses `Ok` and `Err(NonZeroU32)` rather than plain type like `u32`. And I don't know how to enum an `NonZeroU32`

Reply:
A fancy enum with an `Ok` variant and an `Err(NonZeroU32)`  is just a u32, so sending it to the network should be fine. It's somewhat mentioned [here](https://doc.rust-lang.org/std/option/index.html#representation). In some rare cases, the memory layout is not optimized somehow, so we add a bunch of `static_asserts` for that.

I am now more tending to replace `TransportStatus` with a more general `Status` (or `RuntimeStatus`) when it is bubbled up to the user, since there might be other reasons indicating an error, and transport layer is just one source of the errors. We can also extend the size of `RuntimeStatus`. It does not have to be exactly 4 bytes. It just needs to be predictable-sized.

So maybe something like
```rust
enum RuntimeStatus {
    Success,
    /// The underlying transport error code
    Transport(NonZeroU32),
    /// Other plugin specific non-success status
    Other(StatusCode),
}
```

Reply:
I agree. But that seems to require a lot code change, maybe we should merge this PR first, and start a new branch and issue to track on the status code?

Reply:
Sounds great